### PR TITLE
Add Serial support for Windows OS

### PR DESCRIPTION
**Pull request by doganmurat**
Fix https://github.com/MadLittleMods/node-usb-detection/issues/36

Add Serial support for Windows OS

---

Alternative to https://github.com/MadLittleMods/node-usb-detection/pull/50

### CODE REVIEW

**Comment by doganmurat** [comment]
Hi,
@jasondana & @MadLittleMods could you please try this code.
Regards,

File: src/detection_win.cpp [hunk 1]
```diff
@@ -423,6 +424,21 @@ void ExtractDeviceInfo(HDEVINFO hDevInfo, SP_DEVINFO_DATA* pspDevInfoData, TCHAR
 		// Use this to extract VID / PID
 		extractVidPid(buf, resultItem);
 	}
+
+	DWORD dwCap = 0x0;
+	if (DllSetupDiGetDeviceRegistryProperty(hDevInfo, pspDevInfoData, SPDRP_CAPABILITIES, &DataT, (PBYTE)&dwCap, 4, &nSize)) {
+		// If device has a CM_DEVCAP_UNIQUEID capability, then we assume it has a serial number
+		if ((dwCap & CM_DEVCAP_UNIQUEID) == CM_DEVCAP_UNIQUEID) {
+			if (DllSetupDiGetDeviceInstanceId(hDevInfo, pspDevInfoData, buf, buffSize, &nSize)) {
+				// extract Serial Number
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Let's add a comment about what we are trying to parse

```
// Format: <device-ID>\<instance-specific-ID>
//
// Ex. `USB\VID_2109&PID_8110\5&376ABA2D&0&21`
//  - `<device-ID>`: `USB\VID_2109&PID_8110`
//  - `<instance-specific-ID>`: `5&376ABA2D&0&21`
//
// [Device instance IDs](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/device-instance-ids) ->
//  - [Device IDs](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/device-ids) -> [Hardware IDs](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/hardware-ids) -> [Device identifier formats](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/device-identifier-formats) -> [Identifiers for USB devices](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/identifiers-for-usb-devices)
//     - [Standard USB Identifiers](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/standard-usb-identifiers)
//     - [Special USB Identifiers](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/special-usb-identifiers)
//  - [Instance specific ID](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/instance-ids) 
```

---

From https://github.com/MadLittleMods/node-usb-detection/pull/50#discussion_r153106778

File: src/detection_win.cpp [hunk 2]
```diff
@@ -423,6 +424,21 @@ void ExtractDeviceInfo(HDEVINFO hDevInfo, SP_DEVINFO_DATA* pspDevInfoData, TCHAR
...
+				string str = buf;
+				size_t t = str.find_last_of("\\\\");
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: We are parsing strings like `USB\VID_046D&PID_C24E\9025267AF90009`

Why are we using a double backslash `\\\\`? It seems to work just fine with `str.find_last_of("\\");`. According to the docs, we should only need to define the backslash once, http://www.cplusplus.com/reference/string/string/find_last_of/ 🤷‍♂️ 

---

Let's rename this variable to `size_t serialNumberIndex = ...` to make this more clear

File: src/detection_win.cpp [hunk 3]
```diff
@@ -423,6 +424,21 @@ void ExtractDeviceInfo(HDEVINFO hDevInfo, SP_DEVINFO_DATA* pspDevInfoData, TCHAR
...
 		extractVidPid(buf, resultItem);
 	}
+
+	DWORD dwCap = 0x0;
+	if (DllSetupDiGetDeviceRegistryProperty(hDevInfo, pspDevInfoData, SPDRP_CAPABILITIES, &DataT, (PBYTE)&dwCap, 4, &nSize)) {
+		// If device has a CM_DEVCAP_UNIQUEID capability, then we assume it has a serial number
+		if ((dwCap & CM_DEVCAP_UNIQUEID) == CM_DEVCAP_UNIQUEID) {
+			if (DllSetupDiGetDeviceInstanceId(hDevInfo, pspDevInfoData, buf, buffSize, &nSize)) {
+				// extract Serial Number
+				string str = buf;
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Let's rename this variable to `string deviceInstanceId = buf;` so it better represents what it holds

File: src/detection_win.cpp [hunk 4]
```diff
@@ -7,6 +7,7 @@
 
 // Include Windows headers
 #include <windows.h>
+#include <Cfgmgr32.h>
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Why is this needed now?

File: src/detection_win.cpp [hunk 5]
```diff
@@ -423,6 +424,21 @@ void ExtractDeviceInfo(HDEVINFO hDevInfo, SP_DEVINFO_DATA* pspDevInfoData, TCHAR
 		// Use this to extract VID / PID
 		extractVidPid(buf, resultItem);
 	}
+
+	DWORD dwCap = 0x0;
+	if (DllSetupDiGetDeviceRegistryProperty(hDevInfo, pspDevInfoData, SPDRP_CAPABILITIES, &DataT, (PBYTE)&dwCap, 4, &nSize)) {
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Why `4` instead of `buffSize` like the other instances we call `DllSetupDiGetDeviceRegistryProperty`?

File: src/detection_win.cpp [hunk 6]
```diff
@@ -423,6 +424,21 @@ void ExtractDeviceInfo(HDEVINFO hDevInfo, SP_DEVINFO_DATA* pspDevInfoData, TCHAR
 		// Use this to extract VID / PID
 		extractVidPid(buf, resultItem);
 	}
+
+	DWORD dwCap = 0x0;
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: `dwCapabilities`

File: src/detection_win.cpp [hunk 7]
```diff
@@ -423,6 +424,21 @@ void ExtractDeviceInfo(HDEVINFO hDevInfo, SP_DEVINFO_DATA* pspDevInfoData, TCHAR
...
+	if (DllSetupDiGetDeviceRegistryProperty(hDevInfo, pspDevInfoData, SPDRP_CAPABILITIES, &DataT, (PBYTE)&dwCap, 4, &nSize)) {
+		// If device has a CM_DEVCAP_UNIQUEID capability, then we assume it has a serial number
+		if ((dwCap & CM_DEVCAP_UNIQUEID) == CM_DEVCAP_UNIQUEID) {
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Just for reference, I found this other usage

 - https://github.com/Fedict/eid-edt/blob/79dd4cfe7373714dbfbb645ea63e370e357e9cad/Win/src/devices.cpp#L292

File: src/detection_win.cpp (hunk 1 above)
Comment (doganmurat) [reply src/detection_win.cpp]: I am OK with that

File: src/detection_win.cpp (hunk 2 above)
Comment (doganmurat) [reply src/detection_win.cpp]: Why double backslash? 
I have no idea :) just miswritten.

File: src/detection_win.cpp (hunk 3 above)
Comment (doganmurat) [reply src/detection_win.cpp]: I am OK with that

File: src/detection_win.cpp (hunk 4 above)
Comment (doganmurat) [reply src/detection_win.cpp]: CM_DEVCAP_UNIQUEID is defined in that header file, so we have to include it.

File: src/detection_win.cpp (hunk 6 above)
Comment (doganmurat) [reply src/detection_win.cpp]: I am OK with that

File: src/detection_win.cpp (hunk 5 above)
Comment (doganmurat) [reply src/detection_win.cpp]: Because we are trying to get a DWORD, so it is exactly 4.

File: src/detection_win.cpp (hunk 7 above)
Comment (doganmurat) [reply src/detection_win.cpp]: I think our way is better & more clear.

**Comment by doganmurat** [comment]
Hi,
I reviewed my code & tested again.

File: src/detection_win.cpp (hunk 5 above)
Comment (MadLittleMods) [reply src/detection_win.cpp]: Can we use `sizeOf(DWORD)` or something similar?

File: src/detection_win.cpp [hunk 8]
```diff
@@ -7,6 +7,7 @@
 
 // Include Windows headers
 #include <windows.h>
+#include <Cfgmgr32.h>		// CM_DEVCAP_UNIQUEID
```
Comment (MadLittleMods) [inline src/detection_win.cpp]: Let's put the comment above 

```
// Include `CM_DEVCAP_UNIQUEID`
```

File: src/detection_win.cpp (hunk 4 above)
Comment (MadLittleMods) [reply src/detection_win.cpp]: I don't see it listed here

 - https://msdn.microsoft.com/en-us/library/windows/hardware/mt813619%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
 - https://github.com/msysgit/msysgit/blob/e6cc6bf1bba06c7029b2d59aedb0c975a0c4f6fe/mingw/include/ddk/cfgmgr32.h

But the docs seem to indicate it should be :+1:

> SPDRP_CAPABILITIES
> 
> The function retrieves a bitwise OR of the following CM_DEVCAP_Xxx flags in a DWORD. The device capabilities that are represented by these flags correspond to the device capabilities that are represented by the members of the DEVICE_CAPABILITIES structure. The CM_DEVCAP_Xxx constants are defined in Cfgmgr32.h.
>
> https://msdn.microsoft.com/en-us/library/windows/hardware/ff551967(v=vs.85).aspx

**Comment by MadLittleMods** [comment]
@doganmurat Still a question here, https://github.com/MadLittleMods/node-usb-detection/pull/62#discussion_r191127711

File: src/detection_win.cpp (hunk 5 above)
Comment (doganmurat) [reply src/detection_win.cpp]: Actually it is defined as;
typedef unsigned long       DWORD;
And will not change in the future.
But no problem we can use sizeof(DWORD).

Why do we use buffSize for all other instances? What's different about this one?
In other instances we are reading data to buf variable. In this instance we are reading data to dwCapabilities which is a DWORD.

**Comment by doganmurat** [comment]
Hi @MadLittleMods ,
Any other comment or question?

**Comment by doganmurat** [comment]
Hi,
You mean header file comment?
It is already in commit.

**Comment by MadLittleMods** [comment]
@doganmurat I linked the specific `sizeof` comment. We are still using `4` instead of `sizeof(dwCapabilities)`, https://github.com/MadLittleMods/node-usb-detection/pull/62#discussion_r191066307

File: src/detection_win.cpp (hunk 5 above)
Comment (doganmurat) [reply src/detection_win.cpp]: Done.

**Comment by doganmurat** [comment]
Hi,
Commit 13a016699f2f5cf8a7299b0cde75a8cbfc1e0dd2 changes code to sizeof(DWORD), and
last commit 3ea0c32332665df788c378c43d01ecb2e643b73b changes code to sizeof(dwCapabilities).

Please check it.

**Comment by doganmurat** [comment]
Hi @MadLittleMods ,

Good news! :v:
Thank you.

**Comment by MadLittleMods** [comment]
@doganmurat v3.1.0 published to npm 🚀, https://www.npmjs.com/package/usb-detection
